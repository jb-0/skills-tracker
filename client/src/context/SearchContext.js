/**
  * This context is used to pass state to the numerous components and events (such as save) that
  * make up a search. This also means a user can leave the search screen to view their profile
  * without losing search information.
  */
import React, { createContext, useEffect, useState  } from 'react';
import getPermittedTerms from '../services/getPermittedTerms';

export const SearchContext = createContext();

export const SearchProvider = (props) => {
  // Store a number of variables in state, firstly the input from the search text box, second the
  // terms that have been added so far and finally the suggested search terms
  const [search, setSearch] = useState({
    searchInputText: '',
    searchTerms: [],
    suggestedTerms: [],
    submittedSearchTerms: [],
    location: 'London',
    submittedLocation: '',
    permittedTerms: {locations: [], skills: []},
    handleDropDownSelectUpdates,
    handleSearchButtonClick
  });

  useEffect(() => {
    async function setPermittedTerms() {
      const permittedTerms = await getPermittedTerms();

      setSearch((previousValues) => {
        return { ...previousValues, permittedTerms };
      });
    }
    
    setPermittedTerms()
  }, []);

  function handleDropDownSelectUpdates(event) {
    const target = event.target.name;
    const value = event.target.value;

    setSearch((previousValues) => {
      return { ...previousValues, [target]: value };
    });
  }

  // When a user hits search ensure the location and searchTerms are current
  function handleSearchButtonClick() {
    setSearch((previousValues) => {
      return {
        ...previousValues,
        submittedSearchTerms: [...previousValues.searchTerms],
        submittedLocation: previousValues.location,
      };
    });
  }
  
  return <SearchContext.Provider value={ [search, setSearch] }>{props.children}</SearchContext.Provider>;
};
