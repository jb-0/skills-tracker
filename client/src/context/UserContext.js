/**
 * The user context is used to carry state that determines whether or not a user details such as
 * whether or not a user is currently authenticated
 */
import React, { createContext, useState } from 'react';
import isAuthenticated from '../services/auth';

export const UserContext = createContext();

export const UserProvider = (props) => {
  // Manage state for two items, firstly checkingAuthentication which is indicative of whether
  // or not LOADING should be displayed, once this is false then whether or not a user is auth'd
  // is known and the user hits the router. This also supplies the function that allows state
  // to be re-confirmed
  const [userState, setUserState] = useState({
    checkingAuthentication: true,
    authenticated: false,
    isAuthenticated,
  });

  return (
    <UserContext.Provider value={[userState, setUserState]}>
      {props.children}
    </UserContext.Provider>
  );
};
