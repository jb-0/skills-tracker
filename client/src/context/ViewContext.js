/**
  * Being able to get the screen dimensions in any component is useful for conditionally rendering
  * based on device type as well as changing class names for styling purpose. This context is used
  * to provide this dimension information as well as classify whether the width is mobile or desktop
  */
import React, { createContext, useState, useLayoutEffect } from 'react';

// Determine device type based on width
function sizeName(width) {
  if (width <= 480) return 'Mobile'
  if (width <= 770) return 'Tablet'
  return 'Desktop'
}

// Set initial window sized on creating context
export const ViewContext = createContext({width: window.innerWidth, 
  height: window.innerHeight, 
  device: sizeName(window.innerWidth)});

export const ViewProvider = (props) => {
  const [size, setSize] = useState();

  useLayoutEffect(() => {
    function updateSize() {
      setSize({ 
        width: window.innerWidth, 
        height: window.innerHeight, 
        device: sizeName(window.innerWidth)
      });
    }

    window.addEventListener('resize', updateSize);
    updateSize();

    return () => window.removeEventListener('resize', updateSize);
  }, []);
  return <ViewContext.Provider value={ size }>{props.children}</ViewContext.Provider>;
};
