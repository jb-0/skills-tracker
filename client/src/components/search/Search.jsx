import React, { useContext, useEffect, useState } from 'react';
import SearchBox from './SearchBox';
import SearchSuggestion from './SearchSuggestion';
import SearchTerms from './SearchTerms';
import SearchResults from './SearchResults';
import DropdownSelect from '../common/DropdownSelect';
import SearchSaveButton from './SearchSaveButton';
import Button from '../common/Button';
import { SearchContext } from '../../context/SearchContext';
import { SearchForm, SearchInputs } from './Search.Styles';

function SearchContainer() {
  const [search]  = useContext(SearchContext);

  // Return the various elements that make up the search view and pass the necessary props
  return (
    <SearchForm>
      {/* While props are used for common components, search functionality uses the SearchContext 
      and as a result no props are passed down */}
      {/* The dropdown selection allows the user to choose a city/location */}
      <SearchInputs>
        <DropdownSelect
          name="location"
          selectOptions={search.permittedTerms.locations}
          value={search.location}
          onChange={search.handleDropDownSelectUpdates}
        />

        {/* The search box allows users to start typing a skill */}
        <div>
        <SearchBox />

        {/* Search suggestions are rendered as the user starts typing in the search box */}
        <SearchSuggestion />
        </div>

        {/* The search button triggers a fetch to the backend to return the number of jobs */}
        <Button
          buttonText="Search"
          buttonAction={search.handleSearchButtonClick}
        />
      </SearchInputs>

      {/* When a suggestion is clicked it is rendered as a selected search term  */}
      <SearchTerms />

      {/* As long as the user has select at least one search term and pressed search, the search
      results will be returned and displayed */}
      {search.submittedSearchTerms.length > 0 ? (
        <SearchResults />
      ) : null}

      {/* When a search is submitted by an authenticated user they will have the option to save
      their search */}
      <SearchSaveButton />

      
    </SearchForm>
  );
}

export default SearchContainer;
