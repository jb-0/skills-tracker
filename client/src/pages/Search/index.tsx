import React from 'react';
import { Autocomplete, Box, FormControl, InputLabel, MenuItem, Select, TextField, Typography } from '@mui/material';
import { SearchContext } from '../../context/SearchContext';
import bgImage from '../../images/search.png';
import bgImageSm from '../../images/search-sm.png';
import { getSearch } from '../../api';
import { useQuery } from '@tanstack/react-query';

const Search: React.FunctionComponent = () => {
  const [search] = React.useContext(SearchContext);
  const searchTerms = search.searchTerms.join(' ');
  const { data: searchResults = {} } = useQuery(
    ['search', searchTerms],
    () => getSearch.fn(searchTerms, search.location),
    { enabled: search.searchTerms.length > 0 }
  );

  const containerPadding = 4;
  const containerTopPadding = 6;

  return (
    <Box
      p={containerPadding}
      pt={containerTopPadding}
      minHeight="100%"
      width="100%"
      sx={{
        backgroundImage: { xs: `url(${bgImageSm})`, md: `url(${bgImage})` },
        backgroundPosition: 'center bottom',
        backgroundRepeat: 'no-repeat',
        backgroundSize: 'cover',
      }}
    >
      <Box maxWidth="sm" mx="auto">
        <Typography variant="h3">Search for your skillset</Typography>
        <Typography>
          Select a location and begin typing a skill, click a skill from the predefined list to add it to your search.
          Finally click the search button to start searching.
        </Typography>
        <Box component="form" mt={4} display="flex" flexDirection="column" alignItems="flex-start" gap={2}>
          {/* Location */}
          <FormControl>
            <InputLabel variant="standard" id="location-label">
              Location
            </InputLabel>
            <Select
              labelId="location-label"
              label="location"
              variant="standard"
              name="location"
              value={search.location}
              onChange={search.handleDropDownSelectUpdates}
            >
              {search.permittedTerms.locations.map((location: string) => {
                return (
                  <MenuItem key={`location_option_${location}`} value={location}>
                    {location}
                  </MenuItem>
                );
              })}
            </Select>
          </FormControl>

          {/* Skills */}
          <Autocomplete
            multiple
            limitTags={4}
            options={search.permittedTerms.skills}
            defaultValue={[]}
            value={search.searchTerms}
            onChange={(_event: React.SyntheticEvent, value: string[]) => search.updateSkills(value)}
            renderInput={(params) => <TextField variant="standard" {...params} label="Skills" />}
            sx={{ width: '100%' }}
          />

          {/* Results */}
          {search.searchTerms.length > 0 && searchResults?.noOfResults ? (
            <Box
              p={1}
              width="fit-content"
              display="flex"
              alignItems="center"
              justifyContent="center"
              sx={{
                borderRadius: '10px',
                background: '#007FFF',
              }}
            >
              <Box>
                <Typography color="common.white">No. of jobs found:</Typography>
                <Typography variant="h3" color="common.white">
                  {searchResults?.noOfResults}
                </Typography>
              </Box>
            </Box>
          ) : null}
        </Box>
      </Box>
    </Box>
  );
};

export default Search;
