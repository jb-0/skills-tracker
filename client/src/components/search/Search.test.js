import '@testing-library/jest-dom/extend-expect';
import React from 'react';
import renderer from 'react-test-renderer';
import { fireEvent, render, screen } from '@testing-library/react';
import { act } from 'react-dom/test-utils';
import userEvent from '@testing-library/user-event';

import Search from './Search';
import { UserProvider } from '../../context/UserContext';
import { SearchProvider } from '../../context/SearchContext';
import { unmountComponentAtNode } from 'react-dom';

describe('Search component', () => {
  let container = null;

  beforeEach(async () => {
    // setup a DOM element as a render target
    container = document.createElement('div');
    document.body.appendChild(container);

    await act(async () => {
      const promise = Promise.resolve()
      render(
        <UserProvider>
          <SearchProvider>
            <Search />
          </SearchProvider>
        </UserProvider>,
        container
      );
      await act(() => promise);
    });
  });

  beforeEach(() => {
    fetch.resetMocks();
  });

  afterEach(() => {
    // cleanup on exiting
    unmountComponentAtNode(container);
    container.remove();
    container = null;
  });

  it('combined search components render correctly with default state', async () => {
    const {act} = renderer;
    const promise = Promise.resolve()

    const tree = renderer
    .create(
      <UserProvider>
        <SearchProvider>
          <Search />
        </SearchProvider>
      </UserProvider>
    );

    await act(() => promise);

    expect(tree.toJSON()).toMatchSnapshot()
  });

  it('clicking search with coventry as the selected location returns results in coventry', async () => {
    // Switch the select from it's default to Coventry
    fireEvent.change(screen.getByTestId("select"), {
      target: { value: "Coventry" },
    });

    // At this point Coventry should only appear once on the page
    expect(screen.queryAllByText(/Coventry/)).toHaveLength(1);

    // Construct a positive mock server response
    fetch.mockResponseOnce(
      JSON.stringify({ noOfResults: 254, msg: 'results found' })
    );

    act(() => {
      // Type desired search term
      fireEvent.change(screen.getByRole('textbox'), {
        target: { value: 'Jav' },
      });
    });
    // Click on the suggested term Javascript to add it to add it to the search terms array
    userEvent.click(screen.getByText('JavaScript'));

    await act(async () => {
      // Click the search button to trigger search
      userEvent.click(screen.getByRole('button'));
    });

    // Having run a search, based on the mock response we should see Coventry twice
    expect(screen.queryAllByText(/Coventry/)).toHaveLength(2);
  });
});
