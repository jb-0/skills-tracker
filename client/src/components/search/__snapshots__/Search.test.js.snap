// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search component combined search components render correctly with default state 1`] = `
<form
  className="sc-gKsewC joogot"
>
  <div
    className="sc-iBPRYJ SwaKl"
  >
    <select
      className="sc-eCssSg iihBUJ"
      data-testid="select"
      id="location"
      name="location"
      onChange={[Function]}
      value="London"
    >
      <option
        value="Coventry"
      >
        Coventry
      </option>
      <option
        value="Essex"
      >
        Essex
      </option>
      <option
        value="London"
      >
        London
      </option>
    </select>
    <div>
      <input
        autoComplete="off"
        className="sc-gsTCUz gmbZjy"
        id="searchInputText"
        name="searchInputText"
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="Start typing a skill.."
        type="text"
        value=""
      />
    </div>
    <button
      className="sc-jSgupP gjKCMd"
      onClick={[Function]}
      type="button"
    >
      Search
    </button>
  </div>
  <div />
  <div />
</form>
`;
